### PR TITLE
add channel last 3d support for maxpool3d on CPU

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98819
* __->__ #97775

### Testing
Single socket (28 cores):

shape | fp32 forward / ms | bf16 forward / ms | fp32 backward / ms  | bf16 backward / ms
-- | -- | -- | -- | --
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: contig | 3.959584 | 5.493402 | 0.557232 | 0.568485
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: CL | 0.815511 | 1.351261 | 5.710506 | 10.57506
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: contig  | 10.63426 | 15.28637 | 2.67656 | 1.71365
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: CL | 2.63570 | 2.05532 | 2.55452 | 2.33923
size: (4, 19, 10, 16, 16), kernel:   3, stride: 1, mem_format: contig | 0.375469 | 0.479748 | 0.066364 | 0.065155
size: (4, 19, 10, 16, 16), kernel:   3, stride: 1, mem_format: CL3d | 0.112197 | 0.112326 | 0.111697 | 0.145364

Single core:

shape | fp32 forward / ms | bf16 forward / ms | fp32 backward / ms | bf16 backward / ms
-- | -- | -- | -- | --
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: contig | 92.16582 | 128.6513 | 6.684325 | 12.21541
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: CL | 10.14318 | 29.80297 | 7.350142 | 11.25323
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: contig | 238.55453 | 331.89967 | 19.694657 | 32.78853
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: CL | 30.17079 | 32.75628 | 22.44543 | 30.17796
size: (4, 19, 10, 16, 16), kernel:   3, stride: 1, mem_format: contig | 7.474389 | 9.937217 | 0.236015 | 0.434229
size: (4, 19, 10, 16, 16), kernel:   3, stride: 1, mem_format: CL3d | 2.318954 | 2.469444 | 0.262125 | 0.401361

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/DilatedMaxPool3d.cpp [hunk 1]
```diff
@@ -1,11 +1,14 @@
+<<<<<<< HEAD
```
Comment (mingfeima) [inline aten/src/ATen/native/DilatedMaxPool3d.cpp:1]: Oops, some rebase error here ~_~

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/DilatedMaxPool3d.cpp (hunk 1 above)
Comment (CaoE) [reply aten/src/ATen/native/DilatedMaxPool3d.cpp]: Fixed the rebase error.

File: aten/src/ATen/native/cpu/MaxPoolKernel.cpp [hunk 2]
```diff
@@ -88,7 +252,94 @@ void cpu_max_pool(
   }
 }
 
-template <typename scalar_t>
+template <typename scalar_t, typename accscalar_t>
+void cpu_max_pool_3d(
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/MaxPoolKernel.cpp:256]: possible to share code between `cpu_max_pool` and `cpu_max_pool_3d` too?

**Review by jgong5 — COMMENTED** [review]
Any performance numbers on the channels last 3d optimiziation and bfloat16?

File: test/nn/test_pooling.py [hunk 3]
```diff
@@ -914,21 +914,27 @@ def helper(n, c, h, w, d, kernel_size, stride=None):
         helper(4, 8, 8, 8, 8, 7)
         helper(4, 8, 8, 8, 8, (5, 6, 7))
         helper(1, 8, 8, 8, 8, (5, 6, 7))
-        helper(0, 6, 12, 13, 14, (5, 6, 7))
+        if device != "cpu":
+            helper(0, 6, 12, 13, 14, (5, 6, 7))
```
[the remark this answers is not in the document]
Comment (CaoE) [reply test/nn/test_pooling.py:935]: When dims of input is 4, maxpool3d on CUDA will unsqueeze input at dim 0 and check if it is in channels last 3d. If yes, unsqueeze output and indices, and resize them to channels last 3d too, then squeeze output and indices. The behavior on CPU is different.  Maxpool3d on CPU uses input.suggest_memory_format() to decide whether to go channels last 3d pass. For such case input.suggest_memory_format() is contiguous.

[the remark this answers is not in the document]
Comment (jgong5) [reply test/nn/test_pooling.py:935]: For such a case, input would be suggested as channels first on CPU but channels last contiguous? Why CPU implementation has to be different from CUDA?

**Comment by CaoE** [comment]
> Any performance numbers on the channels last 3d optimiziation and bfloat16?

Will provide performance numbers later.

**Review by jgong5 — APPROVED** [review]

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/native/cpu/MaxPoolKernel.cpp [hunk 4]
```diff
@@ -564,13 +564,36 @@ void max_pool2d_kernel_impl(
...
+    // work around buggy behavior of suggest_memory_format here where
+    // suggested format of unsqueezed tensor is contiguous while it is
+    // really only contiguous in ChannelsLast3d
+    if ((!input_cl_check.is_contiguous()) &&
+                     input_cl_check.is_contiguous(at::MemoryFormat::ChannelsLast3d)) {
+      DimVector out_sizes(output.sizes().begin(), output.sizes().end());
+      out_sizes.insert(out_sizes.begin(), 1);
+      output.resize_(out_sizes, at::MemoryFormat::ChannelsLast3d);
+      DimVector indices_sizes(indices.sizes().begin(), indices.sizes().end());
+      indices_sizes.insert(indices_sizes.begin(), 1);
```
[the remark this answers is not in the document]
Comment (CaoE) [reply aten/src/ATen/native/cpu/MaxPoolKernel.cpp:587]: Added TORCH_CHECK.

**Comment by CaoE** [comment]
@ngimel Could you please review this PR ? Thank you.

**Comment by CaoE** [comment]
@Skylion007 Could you please review this PR ? Thank you.

**Comment by CaoE** [comment]
@cpuhrsch  Could you please review this PR ? Thank you.

**Comment by CaoE** [comment]
@mikaylagawarecki Could you please review this PR ? Thank you.

**Review by mikaylagawarecki — COMMENTED** [review]
Hey @CaoE as a broader comment

Following up from our previous discussion, is https://github.com/pytorch/pytorch/pull/106104 ready for review? Happy to review it if it is (and would prefer that we use that testing before landing the rest of these PRs :)

Separately, we must be wary that the `test_memory_format` tests for ModuleInfos has been skipped rather than xfailed for quite a few ops, as we add channels_last_3d support for more of them, could we unskip these tests as well please e.g. https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_modules.py#L2555

**Comment by CaoE** [comment]
@mikaylagawarecki Could you please review this PR ? https://github.com/pytorch/pytorch/pull/106104 is ready for review. Thank you.

**Comment by CaoE** [comment]
@mikaylagawarecki  Can we merge this PR before 2.1 branch if possible ?

**Review by mikaylagawarecki — APPROVED** [review]
Thanks for using the ModuleInfo testing

**Pull request closed without merging**